beacon-chain/core/altair/epoch_precompute.go

```diff
@@ -0,0 +1,229 @@
+	prevEpoch := helpers.PrevEpoch(state)
+
+	inactivityScores, err := state.InactivityScores()
+	if err != nil {
+		return nil, nil, err
+	}
+
+	// This shouldn't happen with a correct beacon state,
+	// but rather be safe to defend against index out of bound panics.
+	if state.NumValidators() > len(inactivityScores) {
+		return nil, nil, errors.New("num of validators can't be greater than length of inactivity scores")
+	}
+	if err := state.ReadFromEveryValidator(func(idx int, val iface.ReadOnlyValidator) error {
+		// Was validator withdrawable or slashed
+		withdrawable := prevEpoch+1 >= val.WithdrawableEpoch()
+		pVal := &precompute.Validator{
+			IsSlashed:                    val.Slashed(),
+			IsWithdrawableCurrentEpoch:   withdrawable,
+			CurrentEpochEffectiveBalance: val.EffectiveBalance(),
+			InactivityScore:              inactivityScores[idx],
+		}
+		// Validator active current epoch
+		if helpers.IsActiveValidatorUsingTrie(val, helpers.CurrentEpoch(state)) {
+			pVal.IsActiveCurrentEpoch = true
+			bal.ActiveCurrentEpoch += val.EffectiveBalance()
+		}
+		// Validator active previous epoch
+		if helpers.IsActiveValidatorUsingTrie(val, prevEpoch) {
+			pVal.IsActivePrevEpoch = true
+			bal.ActivePrevEpoch += val.EffectiveBalance()
+		}
+
+		pValidators[idx] = pVal
+		return nil
+	}); err != nil {
+		return nil, nil, errors.Wrap(err, "could not initialize epoch validator")
+	}
+	return pValidators, bal, nil
+}
+
+// ProcessEpochParticipation processes the epoch participation in state and update individual validator's pre computes,
+// it also tracks and updates epoch attesting balances.
+func ProcessEpochParticipation(
+	ctx context.Context,
+	state iface.BeaconState,
+	bal *precompute.Balance,
+	vals []*precompute.Validator,
+) ([]*precompute.Validator, *precompute.Balance, error) {
+	ctx, span := trace.StartSpan(ctx, "altair.ProcessEpochParticipation")
+	defer span.End()
+
+	cp, err := state.CurrentEpochParticipation()
+	if err != nil {
+		return nil, nil, err
+	}
+	for i, b := range cp {
+		if HasValidatorFlag(b, params.BeaconConfig().TimelyTargetFlagIndex) {
+			vals[i].IsCurrentEpochTargetAttester = true
+		}
+	}
+	pp, err := state.PreviousEpochParticipation()
+	if err != nil {
+		return nil, nil, err
+	}
+	for i, b := range pp {
+		if HasValidatorFlag(b, params.BeaconConfig().TimelySourceFlagIndex) {
+			vals[i].IsPrevEpochAttester = true
+		}
+		if HasValidatorFlag(b, params.BeaconConfig().TimelyTargetFlagIndex) {
+			vals[i].IsPrevEpochTargetAttester = true
+		}
+		if HasValidatorFlag(b, params.BeaconConfig().TimelyHeadFlagIndex) {
+			vals[i].IsPrevEpochHeadAttester = true
+		}
+	}
+	bal = precompute.UpdateBalance(vals, bal)
+	return vals, bal, nil
+}
+
+// ProcessRewardsAndPenaltiesPrecompute processes the rewards and penalties of individual validator.
+// This is an optimized version by passing in precomputed validator attesting records and and total epoch balances.
+func ProcessRewardsAndPenaltiesPrecompute(
```

Comment:
Can you explain what is the relationship between this function and the two functions, `InitializeEpochValidators` and `ProcessEpochParticipation`? I assume this function will use return values from the other ones, right? It is unclear to me why both functions return `[]*precompute.Validator, *precompute.Balance`. Surely all four returned values can't be passed into `ProcessRewardsAndPenaltiesPrecompute`.

Reply:
They will be used in when we do general epoch transition. You will see in the next PR or take a look at current phase 0 epoch transition code